pandas/core/strings.py

```diff
@@ -1737,10 +1743,71 @@ def cons_row(x):
                 return cons(result, name=name, index=index)
 
     @copy(str_cat)
-    def cat(self, others=None, sep=None, na_rep=None):
+    def cat(self, others=None, sep=None, na_rep=None, align=None):
+        from pandas.core.series import Series
+        from pandas.core.frame import DataFrame
+        if align is None and (isinstance(others, Series)
+                              or (isinstance(others, list)
+                                  and any(isinstance(x, Series)
+                                          for x in others))):
+            align = False
+            warnings.warn("A future version of pandas will perform alignment "
+                          "when others is a series (or a list containing a"
+                          "series). To disable alignment (the previous "
+                          "behavior) and silence this warning, pass "
+                          "'align=False'. To enable alignment (the future "
+                          "behavior) and silence this warning, pass "
+                          "'align=True'", FutureWarning, stacklevel=2)
+
         data = self._orig if self._is_categorical else self._data
-        result = str_cat(data, others=others, sep=sep, na_rep=na_rep)
-        return self._wrap_result(result, use_codes=(not self._is_categorical))
+        if align and isinstance(others, Series):
+            # str_cat deals with arrays only
+            data, others = data.align(others, join='outer')
```

Comment:
I think we should actually align using a left join. The result of `s.str.cat(others)` should always preserve the shape and index of `s` IMO.

Current behaviour:

```
In [19]: s = pd.Series(['a', 'b'])

In [20]: s.str.cat(pd.Series(['a', 'b'], index=[1, 2]), align=True)
Out[20]: 
0    NaN
1     ba
2    NaN
dtype: object
```

Reply:
That's how I started out (see OP), but it would be inconsistent with how index-alignment is handled elsewhere. Being consistent in that trumps shape preservation, IMHO. ``str.cat`` is already special anyway, in that it is the only str-method that allows other Series as input.

```
In [0]: s = pd.Series(['a', 'b'])
In [1]: t = pd.Series(['a', 'b'], index=[1, 2])
In [2]: s.astype(bool) & t.astype(bool)
Out[2]:
0    False
1     True
2    False
dtype: bool
In [3]: s.str.cat(t, align=True)
Out[2]: 
0    NaN
1     ba
2    NaN
dtype: object
```

Reply:
I think it depends a lot on the application which behavior is desired. How about exposing a ``join="inner"|"left"|"right"|"outer"`` keyword, with default ``"left"``?

Reply:
A keyword with `join='left'` as the default makes the most sense to me.

Reply:
Added ``join``-keyword with default ``'left'``.